### DISCUSSION TITLE
How to rename the argocd application name? Eg. I put the application name in the test I want to rename the application like test into dev?

### DISCUSSION THREAD

**Discussion by tamilsaran** [argoproj/argo-cd#8305] [Ideas]
How to rename the argocd application name? Eg. I put the application name in the test I want to rename the application like test into dev?

**Comment by tamilsaran** [comment]
Is it possible to without the deletion process to rename the application

**Comment by chetan-rns** [comment]
Hi @tamilsaran, Argo CD Application is a Kubernetes custom resource, and AFAIK the name metadata in Kubernetes is immutable. I think it's not possible to rename an existing resource. We could export the old resource, rename and create a new one, and delete the old resource.

**Reply by tuusberg** [reply]
@chetan-rns thank you for your reply. Could you expand on "export the old resource"? I am trying to rename an application (removing a `-v2` suffix) and would like to try what you suggested above.

**Reply by ManneW** [reply]
Since the Argo CD Application is a resource in Kubernetes, you can read it using `kubectl` and similar tools.

To list the applications, you could do something like:

```
$ kubectl get applications.argoproj.io -n argocd
```

(given you've installed Argo CD in the `argocd` namespace)

To get the full yaml configuration for a specific application:

```
kubectl get applications.argoproj.io/<application-name> -n argocd -o yaml
```

You can then save this to a file (either by redirecting the output to a file in your terminal or by copy-pasting it), perform the desired modifications and finally apply the updated manifest:

```
$ kubectl apply -f <file-with-updated-application.yaml> -n argocd
```

**Comment by antoinemartin** [comment]
Hello,

Maybe a little late to the discussion, but as I was trying to do the same thing and google brought me here.

I'm adding  what I've done in this thread in case it may be useful. 

What I did first is disable pruning, self-heal and auto-sync on the old app. I then added the app with new name. Argo showed me a "shared resource warning" on the old app for all the application resources (my app is an application of applications).

I then deleted the old application without cascading and ended in the state I was looking for.

**Reply by harshsuccessive** [reply]
but did it kept your old revisions as well ? in case of roll back

**Reply by msobkowiak-olx** [reply]
ArgoCD does not use the "secrets" unlike helm to track the versions as these will be stored in ArgoCD CRDs related to the app. So technically yes - you will lose it.

As @antoinemartin mentioned - you can just delete the app with "non-cascading" option and redeploy with the new name. As long as your app's resources preserve the naming, namespace and other configuration it will work (as assignment to the ArgoCD app is done by an annotation injected to the templates) - in fact this is what we have been using to migrate from helm to Argo with no disruption.

**Comment by msobkowiak-olx** [comment]
Note to the community - technically application "name" does not have to be the "ID", the display name on the UI can be rendered from separate field - so it CAN be supported (although not sure if it will be intuitive because of the metadata field name). As you mentioned the IDs (in this case the metadata "name" field still) are immutable and cannot be changed, however if there is a need to dynamically change the names a tagging-like mechanism should be used (technically the CRD name in that case can even be an UUID).

As I mentioned in my other comment and as @antoinemartin did it is still possible to "rename" the app by destroying it with non-cascading option and just re-creating with a new name.

**Reply by fernandezcuesta** [reply]
IMHO that would be a nice to have, for example with something explicit like a "nameOverride" label/annotation. In my case, I've many similar applications belonging to different projects, which should ideally have the same (display) name but cannot.

**Comment by chzar** [comment]
If anyone is still following this discussion - I have an open PR that implements a UI display name override via an annotation: https://github.com/argoproj/argo-cd/pull/26695

**Comment by ahockersten** [comment]
If your `Application` is controlled by an `ApplicationSet`, the syncPolicy [`preserveResourcesOnDeletion: true`](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Controlling-Resource-Modification/#prevent-an-applications-child-resources-from-being-deleted-when-the-parent-application-is-deleted) can be very useful when renaming an `Application` or `ApplicationSet`. If it is set (**before** the rename), you can safely rename (i.e. recreate) the `Application` or `ApplicationSet` without any resources being destroyed in all the situations I have come across.
